src/compiler/checker.ts

```diff
@@ -43474,6 +43481,31 @@ namespace ts {
             return resolveName(location, reference.escapedText, SymbolFlags.Value | SymbolFlags.ExportValue | SymbolFlags.Alias, /*nodeNotFoundMessage*/ undefined, /*nameArg*/ undefined, /*isUse*/ true);
         }
 
+        /**
+         * Get either a value-meaning symbol or an alias symbol.
+         * Unlike `getReferencedValueSymbol`, if the cached resolved symbol is the unknown symbol,
+         * we call `resolveName` to find a symbol.
+         * This is because when caching the resolved symbol, we only consider value symbols, but here
+         * we want to also get an alias symbol if one exists.
+         */
+        function getReferencedValueOrAliasSymbol(reference: Identifier): Symbol | undefined {
+            const resolvedSymbol = getNodeLinks(reference).resolvedSymbol;
+            if (resolvedSymbol && resolvedSymbol !== unknownSymbol) {
+                return resolvedSymbol;
+            }
+
+            return resolveName(

```

Comment:
Since we're calling `resolveName` if the cache returns `unknownSymbol`, there could be a perf increase in emit time for JS files. I'm not sure if that is a big concern, or how to measure it, or even how to address it, so I'm looking for opinions here.

Reply:
Can this new call happen in any scenario that wasn't broken before your change?

Reply:
If I understand the change correctly (dubious), it sounds like the worst case is a file containing only unused imports?  If so, could you make such a file with the top 100 npm packages and then measure before and after tsc time?

Reply:
Not unused imports, but identifiers that would resolve to unknown symbol in the cache before, which I think means identifiers that either don't have a value meaning (I guess in JS that would be imports we think refer to types), or identifiers that we couldn't resolve (i.e. imports we couldn't resolve).

By file with the top 100 npm packages do you mean a file that imports from the top 100 npm packages?

Reply:
> Not unused imports, but identifiers that would resolve to unknown symbol in the cache before, which I think means identifiers that either don't have a value meaning (I guess in JS that would be imports we think refer to types), or identifiers that we couldn't resolve (i.e. imports we couldn't resolve).

From your tests, it looks like you could just `typeof` each imported symbol?

> By file with the top 100 npm packages do you mean a file that imports from the top 100 npm packages?

Yes, that's what I meant.

Reply:
@gabritto What does your thumbs-up above mean?  "No, it can't happen"?

Reply:
Turns out I was wrong, and we hit the cache for unresolved imports. We don't hit the cache for unresolved identifiers.